### PR TITLE
Separate DNS server and Nginx containers

### LINKED ISSUE #1: [Enhancement]: Separate DNS server and Nginx containers
### Checked Existing

- [X] I have checked the repository for duplicate issues.

### What enhancement would you like to see?

I think it would make more sense to run the DNS server and Nginx in separate containers. For example, someone might have a use for Nginx with support for old TLS versions, but they might not need the DNS server (suppose they are using something like PiHole already and prefer to configure DNS spoofing there). It would also be convenient to provide an example Docker Compose file to build and run both containers.

### Any other details to share? (OPTIONAL)

I would be willing to work on this myself if it is approved.

### PR DESCRIPTION
**Pull request by MatthewL246**
Resolves #1

### Changes:

- Move the DNS server and Nginx configuration to separate containers
- Modify the DNS server Dockerfile to use a standard Node.js template
- Modify the Nginx configuration Dockerfile so that configuration files and certificates need be mounted instead of copied
	- My philosophy is that a Docker image should always build successfully with no manual configuration, and any required user configuration should be set up during the run stage through bind mounts if possible, not the build stage.
	- This is also necessary if this image is ever published, as including certificates in a published image would be bad.
- Add an example Docker Compose file
- Update README documentation and setup steps according to the other changes

[For some reason, if I don't put any text in between these two lists it screws with the formatting]

- [x] I have read and agreed to the [Code of Conduct](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CODE_OF_CONDUCT.md).
- [x] I have read and complied with the [contributing guidelines](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CONTRIBUTING.md).
- [x] What I'm implementing was an [approved issue](../issues?q=is%3Aopen+is%3Aissue+label%3Aapproved).
- [x] I have tested all of my changes.

### CODE REVIEW

**Comment by jonbarrow** [comment]
Going to ping @SuperMarioDaBom and @mrjvs (love you) here since y'all know more about Docker than I do

File: nginx/nginx.example.conf [hunk 1]
```diff
@@ -44,10 +46,10 @@ http {
 			proxy_cache_bypass $http_upgrade;
 		}
 
-		ssl_protocols TLSv1 TLSv1.1;
+		ssl_protocols TLSv1 TLSv1.1 TLSv1.2 TLSv1.3;
```
Comment (jonbarrow) [inline nginx/nginx.example.conf]: Is there a reason these were added? They aren't used by the console, they won't get used

File: compose.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    ports:
+      - 53:53/udp
+    env_file:
+      - ./dns/.env
+
+  nginx:
+    build: ./nginx
+    restart: always
+    ports:
+      - 80:80
```
Comment (jonbarrow) [inline compose.yml:16]: This is specifically made for SSL, does port 80 need to be used?

File: dns/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# SSSL DNS server
+
+Custom DNS server intended to be used in conjunction with [SSSL](https://github.com/PretendoNetwork/SSSL). It redirects
+Nintendo hostnames to an SSSL-powered server using DNS spoofing.
+
+## Config
+
+The only 2 addresses required are for `conntest.nintendowifi.net` and `account.nintendo.net`. These can either be set
+using the default address or explicitly mapping them. Additional addresses may be added using `SSSL_DNS_MAP`.
```
Comment (jonbarrow) [inline dns/README.md]: Why is this 2 lines?

File: dns/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# SSSL DNS server
+
+Custom DNS server intended to be used in conjunction with [SSSL](https://github.com/PretendoNetwork/SSSL). It redirects
+Nintendo hostnames to an SSSL-powered server using DNS spoofing.
```
Comment (jonbarrow) [inline dns/README.md]: Why is this 2 lines?

File: nginx/Dockerfile [hunk 5]
```diff
@@ -1,24 +1,25 @@
...
 ARG nginx_dir="/opt/nginx"
 
 # * Use Ubuntu 18.04 as the build image as it's known-good for TLS 1.0/1.1
 FROM ubuntu:18.04 AS build
 
 # * Install build dependencies
-RUN apt-get update
-RUN apt-get install -y \
+RUN apt-get update \
+	&& apt-get install -y \
```
Comment (jonbarrow) [inline nginx/Dockerfile]: Why was this changed?

File: nginx/Dockerfile [hunk 6]
```diff
@@ -41,12 +42,15 @@ RUN ./configure \
...
 
-# * Use Ubuntu as we need to install several pieces of software
-FROM ubuntu:latest
+# * Use Ubuntu to run Nginx
+FROM ubuntu:latest AS final
+
+RUN apt-get update \
+	&& apt-get install -y \
+	libpcre3 \
+	&& apt-get clean
```
Comment (jonbarrow) [inline nginx/Dockerfile]: Is this required? This is running fine without this, and I'd like to keep the images as light as possible

File: nginx/nginx.default.conf
Comment (jonbarrow) [inline nginx/nginx.default.conf:1]: What's the purpose of this file if the only one that's used is the example file?

**Review by jonbarrow — COMMENTED** [review]
Reviewed while on mobile, forgive any issues or anything I missed

I assume some things are just GitHub's mobile app being weird? For example it's showing several lines on the READMEs as being on multiple lines in the middle of sentences

image

File: dns/README.md (hunk 3 above)
Comment (MatthewL246) [reply dns/README.md]: Hardwrapping. Didn't realize this was a problem, will fix it.

File: nginx/nginx.example.conf (hunk 1 above)
Comment (MatthewL246) [reply nginx/nginx.example.conf]: The main reason is simply that I didn't see any reason to exclude newer protocols.

File: compose.yml (hunk 2 above)
Comment (MatthewL246) [reply compose.yml:16]: I think so, since conntest never uses SSL.

File: nginx/nginx.example.conf (hunk 1 above)
Comment (jonbarrow) [reply nginx/nginx.example.conf]: Well, the reason being that they're completely unused. The only clients that will ever connect here are incapable of using these protocols, so there's no reason to have them in the first place

File: compose.yml (hunk 2 above)
Comment (jonbarrow) [reply compose.yml:16]: This is fair, I forgot about this 👍

File: nginx/Dockerfile (hunk 5 above)
Comment (MatthewL246) [reply nginx/Dockerfile]: According to https://docs.docker.com/develop/develop-images/instructions/#apt-get running them on separate steps can cause caching issues.

File: nginx/Dockerfile (hunk 6 above)
Comment (MatthewL246) [reply nginx/Dockerfile]: Yes, it's required. Otherwise, nginx crashes:

```
nginx: error while loading shared libraries: libpcre.so.3: cannot open shared object file: No such file or directory
```

However, it's only required because you specified `--with-pcre` to the Nginx configure script. I kept it because I assumed you were using the Perl compatible regular expressions in the library, but if it isn't required it can be removed.

File: nginx/nginx.default.conf
Comment (MatthewL246) [reply nginx/nginx.default.conf:1]: The `nginx.default.conf` is the one that's copied into the container as the default config file. It just shows the Nginx welcome page and could be used to test the SSL connections without running anything else.

The `nginx.example.conf` is an example for people who want to redirect traffic to a local Pretendo account server. It's unsuitable to put in the container as a default config because we can't possibly know what the user wants the upstream address to be, and the default should be as simple as possible and work for everyone. Using `127.0.0.1` just causes an infinite loop, and using something else like `host.docker.internal` could cause random undefined behavior for the user.

File: nginx/Dockerfile (hunk 5 above)
Comment (jonbarrow) [reply nginx/Dockerfile]: I was not aware of this 👍 thank you, I did it like that for style reasons but if there's an issue with that then I'm fine changing it

File: nginx/Dockerfile (hunk 6 above)
Comment (jonbarrow) [reply nginx/Dockerfile]: In my tests I was unable to build nginx at all without it, but I'm not using it myself no. We can play around with this

File: nginx/nginx.default.conf
Comment (jonbarrow) [reply nginx/nginx.default.conf:1]: Doesn't nginx ship with a default config for this purpose? Why provide our own?

Comment (MatthewL246) [reply nginx/nginx.default.conf:1]: Their default config doesn't have SSL enabled, so it isn't useful for testing if SSSL is working.

File: nginx/Dockerfile (hunk 6 above)
Comment (MatthewL246) [reply nginx/Dockerfile]: Looks like the HTTP rewrite module is the thing that requires the PCRE. Okay to disable that?

File: nginx/Dockerfile [hunk 7]
```diff
@@ -37,16 +38,14 @@ RUN ./configure \
...
 	--with-openssl-opt="enable-weak-ssl-ciphers" \
-	--with-pcre
+	--without-http_rewrite_module
 RUN make -j$(nproc)
 RUN make install
 
-# * Use Ubuntu as we need to install several pieces of software
-FROM ubuntu:latest
+# * Use Ubuntu to run Nginx
+FROM ubuntu:latest AS final
```
Comment (mrjvs) [inline nginx/Dockerfile]: It's probably best if we don't use `:latest` in images meant for production use. Use a specific version instead (like is done a few lines up)

File: dns/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+ENV NODE_ENV production
+USER node
+
+COPY package.json .
+
+COPY --from=dependencies ${app_dir}/node_modules ${app_dir}/node_modules
+COPY --from=build ${app_dir}/dist ${app_dir}/dist
+
+CMD ["pm2-runtime", "start", "."]
```
Comment (mrjvs) [inline dns/Dockerfile]: Is there still a reason to use pm2 here? As nice as pm2 is to manage nodejs processes on a vps, it's kinda odd to have a process manager (pm2) wrapping our process manager (docker).

It would also prevent docker daemon from knowing if the nodejs app is dead (since pm2 always keeps running)

**Review by mrjvs — CHANGES_REQUESTED** [review]

File: dns/Dockerfile (hunk 8 above)
Comment (mrjvs) [reply dns/Dockerfile]: Just noticed this was already discussed in discord dev channels, whoops. I'll keep the review comment open though as it still applies.

File: nginx/Dockerfile (hunk 7 above)
Comment (MatthewL246) [reply nginx/Dockerfile]: Which version? Latest LTS?

Comment (mrjvs) [reply nginx/Dockerfile]: Same one as used above so the client doesn't have to download yet another Ubuntu layer.

It's just so it doesn't randomly break when a new Ubuntu version releases.

Comment (jonbarrow) [reply nginx/Dockerfile]: The build version uses Ubuntu 18, which has long since left support and is no longer receiving security updates. It's only used as the build image since it's known to ship with support for the older TLS versions and SSL ciphers we need.

Is it really a good idea to actually use it as the running, accessible, image though? Given that it's no longer being supported?

Comment (jonbarrow) [reply nginx/Dockerfile]: I should note that I haven't done extensive testing with other versions as the build image. Newer ones could work, either out of the box or with some extra dependencies. I only used this version since it's known to work, which reduced dev time.

Maybe it's worth exploring if other versions also build correctly, if downloading more images is a concern? Or finding lighter images?

Comment (mrjvs) [reply nginx/Dockerfile]: Are those tls things not needed in runtime?

Comment (mrjvs) [reply nginx/Dockerfile]: Since latest was used as final stage before the refactor. Latest LTS should just work, so Id go with that.

I don't think it's worth the time investment of trying more versions. It doesn't impact performance other than a bit more storage space. We just need the image not to break when an Ubuntu update releases by specifying a fixed version.

Comment (jonbarrow) [reply nginx/Dockerfile]: > Are those tls things not needed in runtime?

They are, but not necessarily by the OS. Just by nginx/OpenSSL. But trying to install nginx normally on Ubuntu 20+ won't have support for the TLS/SSL we need. That's why nginx/OpenSSL is installed and built from source on a system designed to support those versions, and then is moved to the final image.

Ubuntu dropping support for TLS 1.0/1.1 after 18 is the reason we switched away from nginx entirely in favor of CF Tunnels, and why Juxt was using Ubuntu 18 containers for a while before that.

So long as nginx/OpenSSL can support the TLS versions and SSL ciphers we need then it doesn't matter what the OS version is tbh.

Comment (jonbarrow) [reply nginx/Dockerfile]: > Since latest was used as final stage before the refactor. Latest LTS should just work, so Id go with that.
> 
> I don't think it's worth the time investment of trying more versions. It doesn't impact performance other than a bit more storage space. We just need the image not to break when an Ubuntu update releases by specifying a fixed version.

That sounds reasonable. Maybe we should lock it to the current LTS version? I believe that's 22?

Comment (MatthewL246) [reply nginx/Dockerfile]: > it's known to ship with support for the older TLS versions and SSL ciphers we need

Why is this necessary if we compile OpenSSL ourselves?

Comment (MatthewL246) [reply nginx/Dockerfile]: Looks like it works fine with the latest Ubuntu LTS, 24.04.

File: nginx/Dockerfile [hunk 9]
```diff
@@ -37,16 +38,14 @@ RUN ./configure \
 	--with-http_ssl_module \
 	--with-openssl=/tmp/openssl-${openssl_version} \
 	--with-openssl-opt="enable-weak-ssl-ciphers" \
-	--with-pcre
+	--without-http_rewrite_module
 RUN make -j$(nproc)
```
Comment (MatthewL246) [inline nginx/Dockerfile:33]: The Nginx Makefile seems to compile OpenSSL by itself, so we don't actually need to compile OpenSSL here separately.

Currently, we are compiling OpenSSL twice.

**Comment by MatthewL246** [comment]
I can confirm that the server still works after these changes on a Wii U without homebrew loaded.

File: dns/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+	npm ci --omit=dev
+
+
+# * Installing development dependencies and building the application
+FROM base AS build
+
+RUN --mount=type=bind,source=package.json,target=package.json \
+	--mount=type=bind,source=package-lock.json,target=package-lock.json \
+	--mount=type=cache,target=/root/.npm \
+	npm ci
```
Comment (jonbarrow) [inline dns/Dockerfile:27]: Why is this done in 2 steps?

File: nginx/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	--prefix=${nginx_dir} \
+	--with-http_ssl_module \
+	--with-openssl=/tmp/openssl-${openssl_version} \
+	--with-openssl-opt="enable-weak-ssl-ciphers" \
+	--without-http_rewrite_module
+RUN make -j$(nproc)
+RUN make install
+
+
+FROM ubuntu:24.04 AS final
```
Comment (jonbarrow) [inline nginx/Dockerfile:37]: Honestly if it's using the same version, why not just pull the source and build in the final image rather than having a separate build step? Is there a benefit to doing it this way?

Comment (MatthewL246) [reply nginx/Dockerfile:37]: Yes, you get a smaller final image without build dependencies included. `build-essential` is pretty large.

File: dns/Dockerfile (hunk 10 above)
Comment (mrjvs) [reply dns/Dockerfile:27]: its two layers, not two steps. one is for runtime and one has all the dev deps to build the project.
they run in parallel so its still speedy.

its to keep the the final image small and without build-only stuff

(I took so long writing an answer that I was ninja'd)

Comment (MatthewL246) [reply dns/Dockerfile:27]: This builds using the development dependencies and then adds only the production dependencies to the final image. So, the final image doesn't have unnecessary `node_modules` like eslint or the types.

File: nginx/Dockerfile (hunk 11 above)
Comment (mrjvs) [reply nginx/Dockerfile:37]: Benefit is not having build-only dependencies and other misc files around in your final running container. The only thing that actually gets published is the final layer, so you try to keep that as small as possible.

It also makes it easier to switch to a different distro (like the very minimal alpine) later when needed, but only for the runtime.

(I took so long writing an answer that I was ninja'd)

Comment (jonbarrow) [reply nginx/Dockerfile:37]: I was assuming it would just be cleaned up after being built, not left

Comment (MatthewL246) [reply nginx/Dockerfile:37]: That's not how Docker works, though. It's more like Git. When you remove files, the layers from before you removed the files still need to be downloaded and stored, so cleaning up won't help.

**Comment by jonbarrow** [comment]
Lgtm, I'll trust that your tests were accurate and this still works :+1:
